Add Ruby interceptors

splittingred opened the pull request:
This PR adds server and client interceptors to the Ruby gRPC libraries.

Server Interceptors must extend the new `GRPC::ServerInterceptor` class, which extends the `GRPC::Interceptor` class. Server interceptors run methods based on the type of request coming in (e.g. `request_response()`). They _must_ yield the block that is passed to complete the interception chain.

Server Interceptors are _server-centric_, meaning they are set on a server and can be used across any service that server is running. Utilizing the passed in desc/handler, one can more specifically target their logic in their own interceptors. I did this following the pattern in [go-grpc-middleware](https://github.com/grpc-ecosystem/go-grpc-middleware) to keep standardization with that library.

Client Interceptors are _stub-centric_, meaning they are set per-ClientStub (_not_ per call).

Interceptors execute in a FIFO chain, and wrap one another, which allows for outer instrumentation or authentication interceptors to happen before other interceptors, yet still be able to measure timing/access changes that might occur from inner, later interceptors.

gRFC: https://github.com/grpc/proposal/pull/34

----

I'm definitely looking for feedback here - if this is not in line with the general direction/consensus of where we want the Ruby gRPC library to go, I'm happy to change/revoke this PR. 

Thanks!

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-testing commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

apolcyn commented:
jenkins: this is ok to test

apolcyn commented:
Thanks for the contribution! With a first look, I think this looks a really good patch.

+ @mehrdada, who's implementing interceptors in other languages.

splittingred commented:
@apolcyn Thanks! Just an fyi - I went ahead and moved the `intercept` call into `RpcDesc` - as I noted in the description above, I feel like that's a more usable place for it, as it allows interceptors to throw `BadStatus` codes (as well as others) and have the `run_server_method` call properly handle unhappy paths. The prior location (in `RpcServer` would have required interceptors to duplicate the rescue handling in `RpcDesc` if they wanted to maintain functionality; I wasn't crazy about that. This should be a little cleaner.

mehrdada commented:
@splittingred Thanks for the PR! I am wondering how difficult it would be to have client interceptors as well as server ones in a single PR so we can ensure the design applies consistently across both sides? In other languages like Go and Java, generally the client interceptors are registered on a channel (or wrap a channel and return another object, in a decorator-like pattern).

splittingred commented:
@mehrdada Sure. I can take a look - my initial thoughts are that due to the way `GenericService.rpc_stub_class` works, we'd have to wrap the dynamic method definitions there in the intercept call to properly handle it across request types. Did either you or @apolcyn have any thoughts there? Happy to take a stab at it.

mehrdada commented:
@splittingred I'd like to defer to @apolcyn on this one, since I don't grok Ruby per se. IMO, in abstract, language-agnostic terms, I think there should be a way to construct an "InterceptedChannel" from a Channel that you would then create a Stub on top of. This way, you can have a potentially different set of interceptors operating on different Stubs.

apolcyn commented on src/ruby/lib/grpc/generic/rpc_desc.rb:
```diff
@@ -72,18 +72,20 @@ def handle_bidi_streamer(active_call, mth)
-      else  # is a bidi_stream
-        handle_bidi_streamer(active_call, mth)
+    def run_server_method(active_call, mth, interceptors = [])
+      intercept(interceptors, active_call, mth) do
```
I agree that the interceptor should be here - especially for the exception handlers for cleanup/finishing of calls

splittingred replied:
I believe this is what `go` does - each interceptor registers and has different methods that are called on the interceptor based on what type of request it is. We can go this route, if we'd prefer.

apolcyn commented on src/ruby/spec/generic/interceptors_spec.rb:
```diff
@@ -0,0 +1,221 @@
+      let(:interceptor3) { TestServerInterceptor.new }
+
+      before(:each) do
+        @server.interceptors[:test] = interceptor
```
I'm wondering about the use of the hash with interceptors using opaque string/symbol identifiers.

Is this a way to de-register or change already configured interceptors?

splittingred replied:
Yes, and it provides a clean interface to the registry of interceptors; you can easily deregister one as if it were a Hash: `server.interceptors[:test] = nil`, or change it by assigning a new interceptor to the same key. I took this approach with the registry as a Hash is generally the easiest way to ensure uniqueness for interceptors within it, while providing a interface that Ruby developers would be automatically familiar with. Happy to change if we feel that's not a good idea.

apolcyn commented:
Should we add client-side interceptors, I'm not exactly sure what we'd want out of them (e.g., per-message intercepts or per-call intercepts?). 

If we are going to do add them though, then indeed it might make sense to add them in the same PR as server interceptors if they're working together at all. And in that case the `GenericService.rpc_stub_class` seems to me like it has direct access to the input and return values of all of the different RPCs.

But what I'm mostly wondering right now is:

* should interceptors have access to <i>messages</i>, or just RPC's and their "call objects"
* should server-side interceptors have the same access to the `active call` that handlers do - e.g., should they be allowed to do things like sending initial metadata.

A design of interceptors for other languages should be documented soon. It might be useful to wait for that to compare this API to other languages.

splittingred commented:
```
Should we add client-side interceptors, I'm not exactly sure what we'd want out of them (e.g., per-message intercepts or per-call intercepts?).
```

In my experience, we've used them elsewhere to:

* Propagate the request context, which is useful for systems like [linkerd](https://linkerd.iuo)/[envoy](https://lyft.github.io/envoy/)
* Measure client calls to grpc services via [statsd](https://github.com/Etsy/statsd)
* Implement spans for distributed tracing, a la [zipkin](http://zipkin.io/)

> If we are going to do add them though, then indeed it might make sense to add them in the same PR as server interceptors if they're working together at all. And in that case the GenericService.rpc_stub_class seems to me like it has direct access to the input and return values of all of the different RPCs.

That was my initial thought, though I understand that the java (and other) clients use the channel for client interception.

> should interceptors have access to messages, or just RPC's and their "call objects"

IMO, yes - at least at a read-only level, as that can be used to determine stats key routing, auth, dynamic marshalling of domain objects into rpc messages, etc.

> should server-side interceptors have the same access to the active call that handlers do - e.g., should they be allowed to do things like sending initial metadata.

We've used interceptors to inject metadata into the request, which is useful for scenarios I listed above.

mehrdada commented:
@splittingred I think I'm in agreement with all your design points, including at least read-only message access. The one I am implementing in C# looks similar: https://github.com/grpc/grpc/pull/12143 
The primary difference is I realized the equivalent to the Channel object is not exposed in the Ruby API at the moment, so perhaps the registration of interceptors should happen at Stub creation point or somewhere it makes more sense, if the Channel is not going to be publicized.

splittingred commented:
> The primary difference is I realized the equivalent to the Channel object is not exposed in the Ruby API at the moment, so perhaps the registration of interceptors should happen at Stub creation point or somewhere it makes more sense, if the Channel is not going to be publicized.

Yeah, this was my main issue as well (similarly in the PHP client). We could take the NodeJS approach, which is passing it through the client call (though I'm not a fan of having to specify an interceptor per-call), or instead maintain it in a client interceptor registry on the stub itself (my preference).

I took a look at the same - specifically creating an `InterceptedChannel` object, but yes, you're right: there's no exposed `GRPC::Channel` object at the moment, and so I was overriding things in the C-bindings which felt dirty, as well as put this PR at risk of ballooning very quickly.

mehrdada commented:
My preference is for per-stub over per-call.

splittingred commented:
@apolcyn @mehrdada I've added Client Interceptors and streamlined the Server Interceptors to have different handler points for each type of gRPC requests. Mind taking a look?

splittingred commented on src/ruby/lib/grpc/generic/client_stub.rb:
```diff
@@ -149,16 +153,29 @@ def request_response(method, req, marshal, unmarshal,
+        method: method,
+        request: req,
+        call: c,
+        metadata: metadata
```
One thing on this (and other request types) was how much information we wanted to pass to interceptors. For example, do we care about passing the deadline, parent, and credentials options? Should that be encapsulated in a `call_options` hash that gets passed to the Interception Context? Looking for thoughts here.

splittingred commented on src/ruby/lib/grpc/generic/interceptors.rb:
```diff
@@ -0,0 +1,167 @@
+    # @param [Method] method
+    # @param [Hash] metadata
+    #
+    def request_response(*)
```
I'd love to explicitly define the args here, but there are a few upstream problems with Rubocop that causes build failures:

- The gRPC Rubocop version is pretty old (0.30) compared to latest (0.49), and the newer versions have a IgnoreUnusedKeywordArguments flag we could use, as well as inline ignore statements
- Upgrading Rubocop is a pretty large change (last I tried on master it was around 120 Rubocop errors or so) and I didn't want to inflate this already larger PR further

apolcyn replied:
I agree that upgrading RuboCop in this PR would make the scope too large.

I'd be ok with logging the arguments, as a hack for now.

apolcyn commented:
> Yes, and it provides a clean interface to the registry of interceptors; you can easily deregister one as if it were a Hash: server.interceptors[:test] = nil, or change it by assigning a new interceptor to the same key. I took this approach with the registry as a Hash is generally the easiest way to ensure uniqueness for interceptors within it, while providing a interface that Ruby developers would be automatically familiar with. Happy to change if we feel that's not a good idea.

Personally, I would favor an interceptor-registration scheme (on both the client and the server), that only allows pre-configuring interceptors:
* before the server has started, (before `RpcServer#run` is called`), on the server
* upon client stub creation, (maybe within as a parameter to `ClientStub#new`)

... with the interceptor registries of the client and server frozen from that point on.

Mostly I think it would simplify things by not having to think about the race conditions of active server configuration, and to keep the present one-time setup of the `ClientStub` (I think things like future forking support will likely want to take advantage of the `ClientStub`'s current single point of setup).

Still we can take on that requirement if the active registration/de-registration is important

splittingred commented:
> Personally, I would favor an interceptor-registration scheme (on both the client and the server), that only allows pre-configuring interceptors:
> before the server has started, (before RpcServer#run is called`), on the server
> upon client stub creation, (maybe within as a parameter to ClientStub#new)
> ... with the interceptor registries of the client and server frozen from that point on.

I think that's fine - I don't see a need for registering/de-registering _post_ call run. And I think through passing an interceptor registry hash into initializers we can allow a dynamic interceptor creation _before_ putting them into the `ClientStub` or `RpcServer` options without running into the issues you described. I've updated the code to add new arguments to allow that. The interceptor registry is no longer accessible post-initialization of either the `ClientStub` or `RpcServer` objects.

> And please let me know your thoughts of such a more constricted API

I think it's great feedback, and I don't forsee it being a problem for the client interceptors. In our experience, the largest use cases for client interceptors have been:

* Automatic metadata insertion _pre-call_ (IOW, unable to send metadata, but instead add)
* Timing outbound requests, which providing any interception ability allows
* Inspecting deadline attributes for logging
* Distributed tracing (similar to timing, arguments mostly are unnecessary, except metadata for propagating tracing keys)

The identifying things that are needed in general for interceptors are the service name being accessed, the RPC method being accessed, and any metadata. I could potentially see the ability to cancel an outbound call in an interceptor being useful (for upstream circuit breaking or resilient hot failover handling), but if that's a problem for now it's something IMO we could consider later.

Thanks @apolcyn - I've updated the branch and it's ready for another look. Appreciate the ongoing dialog here!

apolcyn commented:
Thanks for the updates! I am out today btw but will continue reviewing this next week

apolcyn commented on src/ruby/lib/grpc/generic/interceptor_registry.rb:
```diff
@@ -0,0 +1,121 @@
+    #   registry as
+    # @param [GRPC::Interceptor] interceptor The interceptor instance
+    #
+    def []=(name, interceptor = nil)
```
It looks like the `[]=` method can potentially be bypassed, with bad types accidentally invoked.

AFAICS the features of this class are also offered in `Hash`? 

What do you think if we shrink the `InterceptorRegistry` class here, move the type-checking logic into the `InterceporRegistry#initialize` method, and keep only the `build_context` method. 

The user API would then be to pass in a plain `Hash` to `#new` of this class

splittingred commented:
@apolcyn - Thanks! Done - mind taking a look?

apolcyn commented on src/ruby/lib/grpc/generic/active_call.rb:
```diff
@@ -645,5 +666,9 @@ def initialize(wrapped)
+
+    # InterceptableView further limits access to an ActiveCall's methods
+    # for use in interceptors on the client
+    InterceptableView = view_class(:deadline, :metadata, :trailing_metadata)
```
Whoops, sorry! Realizing that `metadata` and `trailing_metadata` are probably not useful as well, since both only show the metadata/trailers sent from the server, and won't contain anything at the time that the interceptor sees them. Can we remove these too please?

splittingred commented:
@apolcyn Done. ♻️ 🙇

splittingred commented:
@apolcyn @mehrdada Anything else I need to do here? Thanks!

mehrdada commented:
Thanks @splittingred for working on this and @apolcyn for the review so far. I will take a high-level look right away, but will leave the detailed review and approval to Alex.

mehrdada commented on src/ruby/lib/grpc/generic/rpc_server.rb:
```diff
@@ -374,15 +380,19 @@ def loop_handle_server_calls
           end
         rescue Core::CallError, RuntimeError => e
-          # these might happen for various reasonse.  The correct behaviour of
+          # these might happen for various reasons.  The correct behaviour of
```
nit: s/behaviour/behavior

splittingred commented on src/ruby/spec/generic/interceptor_registry_spec.rb:
```diff
@@ -0,0 +1,185 @@
+
```
done

splittingred commented:
@mehrdada Done. Comments addressed. As per the hash ordering, Ruby since 1.9 has maintained Hash ordering by being a doubly-linked list: https://www.igvita.com/2009/02/04/ruby-19-internals-ordered-hash/ - (also in [various spec tests](https://github.com/ruby/ruby/blob/6bc742bc6d0dc6b1777d28484f4694af846bfb79/spec/rubyspec/core/hash/keys_spec.rb#L6) in core ruby)

I've been using a similar paradigm for interceptor ordering (indeed, most Ruby systems use Hash for middleware patterns) and it has maintained FIFO order execution. Happy to change if we're worried about it, but I don't really see Ruby changing that behavior.

mehrdada commented:
@splittingred: Sounds good to me.
@apolcyn: PTAL

apolcyn commented on src/ruby/spec/generic/interceptor_registry_spec.rb:
```diff
@@ -0,0 +1,75 @@
+
+    context 'with an interceptor extending GRPC::ServerInterceptor' do
+      it 'should add the interceptor to the registry' do
+        subject
```
I think this test can be removed, since IIC this functionality no longer used

apolcyn commented on src/ruby/spec/generic/interceptor_registry_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    end
+  end
+
+  describe '.to_h' do
```
same here, I think we can remove this too since no longer used

apolcyn commented on src/ruby/spec/support/helpers.rb:
```diff
@@ -0,0 +1,79 @@
+        @server = RpcServer.new({ poll_period: 1 }.merge(server_opts))
+        @port = @server.add_http2_port('0.0.0.0:0', :this_port_is_insecure)
+        @host = "0.0.0.0:#{@port}"
+        @channel = channel || GRPC::Core::Channel.new(
```
can we get rid of the setting of `@channel` here? IIC this is unused

apolcyn commented on src/ruby/spec/support/helpers.rb:
```diff
@@ -0,0 +1,79 @@
+      def build_insecure_stub(klass, host: nil, opts: nil)
+        host ||= @host
+        opts ||= @client_opts
+        klass.new(host, :this_channel_is_insecure, **opts)
```
since we're creating the stub with the `:this_channel_is_insecure` arg here, can we get rid of the setting the "override channel" above? worried about possible silent effects of that override channel

apolcyn commented on src/ruby/spec/generic/interceptor_registry_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    end
+  end
+
+  describe '.[]' do
```
same here, the hash key->value functionality looks unneeded now

apolcyn commented on src/ruby/spec/support/services.rb:
```diff
@@ -0,0 +1,140 @@
+    requests.each do |r|
+      p "In client interceptor: #{r}"
+    end
+    metadata['foo'] = 'bar'
```
nit: s/bar/bar_from_client_streamer? to further make clear the running of the different RPC types?

apolcyn commented on src/ruby/spec/support/services.rb:
```diff
@@ -0,0 +1,140 @@
+    yield
+  end
+
+  # GRPC.logger.info
```
remove this comment please

apolcyn commented on src/ruby/spec/support/services.rb:
```diff
@@ -0,0 +1,140 @@
+class TestServerInterceptor < GRPC::ServerInterceptor
+  def request_response(request:, call:, method:)
+    p "Received request/response call at method #{method}" \
+      " with request #{request} for call #{call}"
```
similarly to the client interceptors, can we add usage of `merge_metadata_to_send` and `output_metadata[]` method in the server, and test for that metadata/trailing metadata in the client to the interceptors in this `TestServerInterceptor` class?

splittingred commented on src/ruby/pb/grpc/testing/duplicate/echo_duplicate_services_pb.rb:
```diff
@@ -34,6 +34,7 @@ class Service
           self.service_name = 'grpc.testing.duplicate.EchoTestService'
 
           rpc :Echo, Grpc::Testing::EchoRequest, Grpc::Testing::EchoResponse
+          rpc :ResponseStream, Grpc::Testing::EchoRequest, stream(Grpc::Testing::EchoResponse)
```
Adding that was the only way to get the tests to pass:

```
  1) Ping protobuf code generation should have the same content as created by code generation
     Failure/Error: expect(got).to eq(want)

       expected: "# Generated by the protocol buffer compiler.  DO NOT EDIT!\n# Source: src/proto/grpc/testing/duplica...choResponse\n        end\n\n        Stub = Service.rpc_stub_class\n      end\n    end\n  end\nend\n"
            got: "# Generated by the protocol buffer compiler.  DO NOT EDIT!\n# Source: src/proto/grpc/testing/duplica...hoResponse)\n        end\n\n        Stub = Service.rpc_stub_class\n      end\n    end\n  end\nend\n"

       (compared using ==)

       Diff:
       @@ -34,6 +34,7 @@
                  self.service_name = 'grpc.testing.duplicate.EchoTestService'

                  rpc :Echo, Grpc::Testing::EchoRequest, Grpc::Testing::EchoResponse
       +          rpc :ResponseStream, Grpc::Testing::EchoRequest, stream(Grpc::Testing::EchoResponse)
                end

                Stub = Service.rpc_stub_class
```

splittingred commented:
@apolcyn Changes made, and updated the server interceptors to test more of the metadata functionality across the call. PTAL. Thanks!

mehrdada commented:
@splittingred Before we merge this new feature, we need to go through the formal gRFC process (https://github.com/grpc/proposal) and solicit feedback from the community. Would you mind drafting a quick gRFC for your feature? The Node interceptors proposal should serve as a good starting point that you can adapt for this: https://github.com/drobertduke/proposal/blob/nodejs_client_interceptors/L5-NODEJS-CLIENT-INTERCEPTORS.md
Please let me know if that's too much of a burden; I can try and help with that.

splittingred commented:
@apolcyn Done.

@mehrdada Definitely! I'll take a pass at it and push up a PR there. Will ping you two on it when I submit. Thanks!

splittingred commented:
@apolcyn @mehrdada Alright, gRFC submitted. PTAL - Thanks!

@apolcyn Just an FYI - I changed the `InterceptorRegistry` to take in an Array rather than a Hash, since we've hidden its API mostly via the configuration option on the `ClientStub` and `RpcServer`. There wasn't really a reason anymore for keyword storage of interceptors, so I simplified it. I also added a test for ensuring that execution order remains the same when it is added into the `InterceptorRegistry`.

mehrdada commented:
@splittingred I was discussing this PR with @apolcyn right now to see if he has any lingering concerns as the 10-business-day gRFC period is almost over. The only thing that came up was the idea of wrapping the client context in a view so that, for example, the interceptor would not be able to cancel the call or do things that may not be sensible, and in general adding an indirection layer for future proofing purposes. What do you think?

splittingred commented:
@mehrdada I believe that's done through the `InterceptableView` already which only exposes the `deadline` - but maybe you can clarify here?

mehrdada commented:
@splittingred I just looked at the code again and I see only the `deadline` exposed for `ActiveCall`, so you are right. Perhaps I misunderstood @apolcyn, or he was describing a concern that was already addressed and I thought was still remaining.

@apolcyn As far as I am concerned I think we are good to go here. Please confirm or let us know if you have any remaining concerns. Thanks!

splittingred commented:
Hey @apolcyn - anything else you need me to do here?

apolcyn commented:
re-running since most test results are stale

jenkins: test this please

apolcyn commented on src/ruby/lib/grpc/generic/bidi_call.rb:
```diff
@@ -87,23 +87,33 @@ def run_on_client(requests,
 
-      write_loop(replys, is_client: false)
+      write_loop(replies, is_client: false)
+      replies
```
nit: `replies` return value here isn't needed?

apolcyn commented on src/ruby/lib/grpc/generic/rpc_desc.rb:
```diff
@@ -47,43 +47,88 @@ def unmarshal_proc(target)
+          resp,
+          trailing_metadata: active_call.output_metadata
+        )
+        resp
```
nit: similar to above, `resp` return val unneeded?

apolcyn commented on src/ruby/lib/grpc/generic/rpc_desc.rb:
```diff
@@ -47,43 +47,88 @@ def unmarshal_proc(target)
+          resp,
+          trailing_metadata: active_call.output_metadata
+        )
+        resp
```
nit: same as above

apolcyn commented on src/ruby/lib/grpc/generic/rpc_desc.rb:
```diff
@@ -47,43 +47,88 @@ def unmarshal_proc(target)
+        replies = mth.call(req, call)
+        replies.each { |r| active_call.remote_send(r) }
+        send_status(active_call, OK, 'OK', active_call.output_metadata)
+        replies
```
nit: same as above

splittingred commented:
@apolcyn Done! LMK!

mehrdada commented:
Known Failures:
Interop Cloud-to-Prod Tests: https://github.com/grpc/grpc/issues/12618 https://github.com/grpc/grpc/issues/12261 https://github.com/grpc/grpc/issues/10763

kumudt commented on src/ruby/lib/grpc/generic/rpc_desc.rb:
```diff
@@ -47,43 +47,85 @@ def unmarshal_proc(target)
+        request: req
+      ) do
+        resp = mth.call(req, call)
+        active_call.server_unary_response(
```
Hi @splittingred 
Is there any reason why the active_call is sending the response within the yield block? As of now, yield's return value inside a middleware is nil. Looks like the response is sent as soon as the handler method is invoked and the middlewares cannot even read the response.
But, the design document seemed like we can catch the response in the middlewares. Has the design changed keeping the security aspect in mind?
Also, just to make my understanding clear, the server interceptors as of now can only modify the response metadata. Am I right?
